src/rmt.rs

```diff
@@ -0,0 +1,332 @@
+    ///
+    /// Returns None if the resulting ticks from the ns conversion is too high.
+    /// See `PulseTicks` for details.
+    ///
+    /// This function exists on `Writer` because of the internal call to `rmt_get_counter_clock`
+    /// requires a channel which is managed by `Writer`.
+    pub fn pulse_ns(&mut self, pin_state: PinState, ns: u32) -> Result<Option<Pulse>, EspError> {
+        let ticks_per_ns = match &self.ticks_per_ns {
+            None => {
+                let ticks_hz = self.counter_clock()?;
+                let ticks_per_ns = ticks_hz as f32 / 1e9;
+                self.ticks_per_ns = Some(ticks_per_ns);
+                self.ticks_per_ns.as_ref().unwrap()
+            }
+            Some(t) => t,
+        };
+        let ticks = (ticks_per_ns * ns as f32) as u16;
+        match PulseTicks::new(ticks) {
+            Some(ticks) => Ok(Some(Pulse::new(pin_state, ticks))),
+            None => Ok(None),
+        }
+    }
+
+    pub fn add<I>(&mut self, pulses: I) -> Result<(), EspError>
+    where
+        I: IntoIterator<Item = Pulse>,
+    {
+        for pulse in pulses {
+            println!("{:?} {}", &pulse.pin_state, pulse.ticks.0);
+
+            if let Some(item) = self.half_inserted.as_mut() {
+                // SAFETY: We have retrieved this item which is previously populated with the same
+                // union field.
+                let inner = unsafe { &mut item.__bindgen_anon_1.__bindgen_anon_1 };
+
+                inner.set_level1(pulse.pin_state as u32);
+                inner.set_duration1(pulse.ticks.0 as u32);
+
+                self.flush_half_inserted();
+            } else {
+                let mut inner = rmt_item32_t__bindgen_ty_1__bindgen_ty_1::default();
+                inner.set_level0(pulse.pin_state as u32);
+                inner.set_duration0(pulse.ticks.0 as u32);
+                let item = esp_idf_sys::rmt_item32_t {
+                    __bindgen_anon_1: rmt_item32_t__bindgen_ty_1 {
+                        __bindgen_anon_1: inner,
+                    },
+                };
+                self.half_inserted = Some(item);
+            };
+        }
+
+        Ok(())
+    }
+
+    fn flush_half_inserted(&mut self) {
+        if let Some(item) = self.half_inserted {
+            self.items.push(item);
+            self.half_inserted = None;
+        }
+    }
+
+    pub fn clear(&mut self) {
+        self.items.clear();
+        self.half_inserted = None;
+    }
+
+    /// Start sending the pulses.
+    pub fn start(&mut self) -> Result<(), EspError> {
+        self.flush_half_inserted();
+
+        esp!(unsafe {
+            rmt_write_items(
+                self.config.channel as u32,
+                self.items.as_ptr(),
+                self.items.len() as i32,
+                true, // TODO: Blocking.
+            )
+        })
+    }
+
+    pub fn stop(&self) -> Result<(), EspError> {
+        todo!()
+    }
+}
+
+impl Drop for Writer {
```

Comment:
Instead of `Drop` we probably need a `release` method that will stop the `Writer` and return the owned peripherals (the pin and the channel). See [here](https://github.com/esp-rs/esp-idf-hal/blob/master/src/ledc.rs#L199).

Reply:
A release method is needed definitely, but I think having a drop impl that stops the hardware is still a good idea.

Reply:
I don't think it is possible to jave both.